docs/azure/sdk/unit-testing-mocking.md

```diff
@@ -0,0 +1,348 @@
+ms.custom: devx-track-dotnet, engagement-fy23
+ms.date: 07/05/2023
+---
+
+# Unit testing and mocking with the Azure SDK .NET
+
+Unit testing is an important part of a sustainable development process that can improve code quality and prevent regressions or bugs in your app. However, unit testing presents challenges when the code you are testing communicates with an Azure service over a network. Tests that run against live services can experience issues such as latency that slows down test execution, dependencies on code outside of the isolated test, and issues with managing service state and costs every time the test is run. Instead of testing against live Azure services, replace the service clients with mocked or in-memory implementations. This avoids the above issues and lets developers focus on testing their application logic, independent from the network and service.
+
+In this article, we’ll show you how to use the Azure SDK to write great unit tests that isolate your dependencies to make your tests more reliable. First, we’ll show you how many of the Azure SDK building blocks have been designed to be replaced with an in-memory test implementation. Then, we’ll show you how the pieces come together into a fast and reliable unit test. Finally, we’ll provide some design tips to help you design your own classes to better support unit testing.
+
+## Understanding service clients
+
+A service client class is the main entry point for developers in an Azure SDK library and implements most of the logic to communicate with the Azure service. When unit testing service client classes, it’s important to be able to create an instance of the client that behaves as expected without making any network calls.
+
+Each of the Azure SDK clients follows mocking guidelines that allow their behavior to be overridden:
+
+* Each client offers at least one protected constructor to allow inheritance for testing.
+* All public client members are virtual to allow overriding.
+
+> [!NOTE]
+> The code examples in this article use types from the `Azure.Security.KeyVault.Secrets` library for the Azure Key Vault service. The concepts demonstrated in this article also apply to service clients from many other Azure services, such as Azure Storage or Azure Service Bus.
+
+To create a test client instance, inherit from the client type and override methods you are calling in your code with an implementation that returns a set of test objects. Most clients contain both synchronous and asynchronous methods for operations; override only the one your application code is calling.
+
+```csharp
+public class MockSecretClient : SecretClient 
+{ 
+    public MockSecretClient() {}
+
+    public override Response<KeyVaultSecret> GetSecret(string name, string version = null, CancellationToken cancellationToken = default) => ...;
+    public override Task<Response<KeyVaultSecret>> GetSecretAsync(string name, string version = null, CancellationToken cancellationToken = default) => ...;
+
```

Comment:
An empty `.ctor` is redundant, also, lots of folks will Copy this code - and if it doesn't compile that's problematic. Instead of `...` (invalid code), `=> throw new NotImplementedException()`. Also, please try to avoid horizontal scrolling by formatting code for vertical scrolling.

```suggestion
public class MockSecretClient : SecretClient 
{
    public override Response<KeyVaultSecret> GetSecret(
        string name,
        string version = null, 
        CancellationToken cancellationToken = default)
        => throw new NotImplementedException();
        
    public override Task<Response<KeyVaultSecret>> GetSecretAsync(
        string name, 
        string version = null, 
        CancellationToken cancellationToken = default)
        => throw new NotImplementedException();
```